statistics/handle/lru_cache.go

```diff
@@ -0,0 +1,334 @@
+// limitations under the License.
+
+package handle
+
+import (
+	"container/list"
+	"sync"
+
+	"github.com/pingcap/tidb/metrics"
+	"github.com/pingcap/tidb/statistics"
+)
+
+const (
+	typHit    = "hit"
+	typMiss   = "miss"
+	typUpdate = "update"
+	typDel    = "del"
+	typEvict  = "evict"
+	typCopy   = "copy"
+	typTrack  = "track"
+	typTotal  = "total"
+)
+
+// cacheItem wraps Key and Value. It's the value of list.Element.
+type cacheItem struct {
+	key         int64
+	value       *statistics.Table
+	tblMemUsage *statistics.TableMemoryUsage
+}
+
+// internalLRUCache is a simple least recently used cache
+type internalLRUCache struct {
+	sync.RWMutex
+	capacity int64
+	// trackingCost records the tracking memory usage of the elements stored in the internalLRUCache
+	// trackingCost should be kept under capacity by evict policy
+	trackingCost int64
+	// totalCost records the total memory usage of the elements stored in the internalLRUCache
+	totalCost int64
+	elements  map[int64]*list.Element
+	// cache maintains elements in list.
+	// Note that if the element's trackingMemUsage is 0, it will be removed from cache in order to keep cache not too long
+	cache *list.List
+}
+
+// newInternalLRUCache returns internalLRUCache
+func newInternalLRUCache(capacity int64) *internalLRUCache {
+	if capacity < 1 {
+		panic("capacity of LRU Cache should be at least 1.")
+	}
+	return &internalLRUCache{
+		capacity: capacity,
+		elements: make(map[int64]*list.Element),
+		cache:    list.New(),
+	}
+}
+
+// Get tries to find the corresponding value according to the given key.
+func (l *internalLRUCache) Get(key int64) (*statistics.Table, bool) {
+	var hit bool
+	var r *statistics.Table
+	l.Lock()
+	r, hit = l.get(key)
+	l.Unlock()
+	if hit {
+		metrics.StatsCacheLRUCounter.WithLabelValues(typHit).Inc()
+	} else {
+		metrics.StatsCacheLRUCounter.WithLabelValues(typMiss).Inc()
+	}
+	return r, hit
+}
+
+func (l *internalLRUCache) get(key int64) (*statistics.Table, bool) {
+	element, exists := l.elements[key]
+	if !exists {
+		return nil, false
+	}
+	l.cache.MoveToFront(element)
+	return element.Value.(*cacheItem).value, true
+}
+
+// Put puts the (key, value) pair into the LRU Cache.
+func (l *internalLRUCache) Put(key int64, value *statistics.Table) bool {
+	var success bool
+	var trackingCost int64
+	var totalCost int64
+	l.Lock()
+	success = l.put(key, value, value.MemoryUsage(), true)
+	trackingCost = l.trackingCost
+	totalCost = l.totalCost
+	l.Unlock()
+	if success {
+		metrics.StatsCacheLRUCounter.WithLabelValues(typUpdate).Inc()
+		metrics.StatsCacheLRUMemUsage.WithLabelValues(typTrack).Set(float64(trackingCost))
+		metrics.StatsCacheLRUMemUsage.WithLabelValues(typTotal).Set(float64(totalCost))
+	}
+	return success
+}
+
+func (l *internalLRUCache) put(key int64, value *statistics.Table, tblMemUsage *statistics.TableMemoryUsage, tryEvict bool) bool {
+	if l.capacity < tblMemUsage.TotalColTrackingMemUsage() {
+		return false
```

Comment:
Maybe we can drop some column CMSketches to reduce `TotalColTrackingMemUsage` rather than directly return false.

Reply:
updated.